semantic-core/src/Analysis/ImportGraph.hs

```diff
@@ -13,82 +13,101 @@ import           Control.Effect.Fail
 import           Control.Effect.Fresh
 import           Control.Effect.Reader
 import           Control.Effect.State
-import qualified Data.Core as Core
+import           Control.Monad ((>=>))
 import           Data.File
 import           Data.Foldable (fold)
 import           Data.Function (fix)
 import           Data.List.NonEmpty (nonEmpty)
 import           Data.Loc
 import qualified Data.Map as Map
 import           Data.Name
+import           Data.Proxy
 import qualified Data.Set as Set
-import           Data.Term
 import           Data.Text (Text)
 import           Prelude hiding (fail)
 
 type ImportGraph = Map.Map Text (Set.Set Text)
 
-data Value = Value
-  { valueSemi  :: Semi
+data Value term = Value
+  { valueSemi  :: Semi term
   , valueGraph :: ImportGraph
   }
   deriving (Eq, Ord, Show)
 
-instance Semigroup Value where
+instance Semigroup (Value term) where
   Value _ g1 <> Value _ g2 = Value Abstract (Map.unionWith (<>) g1 g2)
 
-instance Monoid Value where
+instance Monoid (Value term) where
   mempty = Value Abstract mempty
 
-data Semi
-  = Closure Loc User (Term (Core.Ann :+: Core.Core) User) User
+data Semi term
+  = Closure Loc User term
   -- FIXME: Bound String values.
   | String Text
   | Abstract
   deriving (Eq, Ord, Show)
 
 
-importGraph :: [File (Term (Core.Ann :+: Core.Core) User)] -> (Heap User Value, [File (Either (Loc, String) Value)])
 importGraph
+  :: (Ord term, Show term)
+  => (forall sig m
+     .  (Carrier sig m, Member (Reader Loc) sig, MonadFail m)
+     => Analysis term User (Value term) m
+     -> (term -> m (Value term))
+     -> (term -> m (Value term))
+     )
+  -> [File term]
+  -> ( Heap User (Value term)
+     , [File (Either (Loc, String) (Value term))]
+     )
+importGraph eval
   = run
   . runFresh
-  . runHeap "__semantic_root"
-  . traverse runFile
+  . runHeap
+  . traverse (runFile eval)
 
-runFile :: ( Carrier sig m
-           , Effect sig
-           , Member Fresh sig
-           , Member (Reader (FrameId User)) sig
-           , Member (State (Heap User Value)) sig
-           )
-        => File (Term (Core.Ann :+: Core.Core) User)
-        -> m (File (Either (Loc, String) Value))
-runFile file = traverse run file
+runFile
+  :: ( Carrier sig m
+     , Effect sig
+     , Member Fresh sig
+     , Member (State (Heap User (Value term))) sig
+     , Ord  term
+     , Show term
+     )
+  => (forall sig m
+     .  (Carrier sig m, Member (Reader Loc) sig, MonadFail m)
+     => Analysis term User (Value term) m
+     -> (term -> m (Value term))
+     -> (term -> m (Value term))
+     )
```

Comment:
Most of the extra lines of code introduced by this PR are reformatted signatures + the evaluator parameters.